compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -511,77 +513,27 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   return success();
 }
 
-/// Returns true if the index map represents a transpose that benefits from
-/// shared mem.
-static bool isSharedMemTranspose(AffineMap indexMap) {
-  if (!indexMap.isEmpty() && indexMap.isPermutation()) {
-    // Ensure that the fasted moving dimension (the last one) is permuted,
-    // Otherwise shared memory promotion will not benefit the operation.
-    if (indexMap.getDimPosition(indexMap.getNumDims() - 1) !=
-        indexMap.getNumDims() - 1) {
-      return true;
-    }
-  }
-  return false;
-}
-
-/// Returns true if the operation is a GenericOp implementing a 2D transpose.
-static bool checkTransposePreconditions(linalg::GenericOp linalgOp) {
-  // Check that the op has at least 2 to 3 parallel loops.
-  if (linalgOp.getNumParallelLoops() < 2 ||
-      linalgOp.getNumParallelLoops() > 3) {
-    return false;
-  }
-
-  // Check that all the iterators are parallel.
-  if (linalgOp.getNumParallelLoops() != linalgOp.getNumLoops()) {
-    return false;
-  }
-
-  // Only transpose static shapes
-  if (linalgOp.hasDynamicShape()) {
-    return false;
-  }
-  return true;
-}
-
 static LogicalResult setTransposeConfig(func::FuncOp entryPoint,
-                                        linalg::GenericOp linalgOp) {
-  if (!checkTransposePreconditions(linalgOp)) {
-    return failure();
-  }
+                                        linalg::LinalgOp linalgOp) {
+  LinalgOpInfo opInfo(linalgOp, sharedMemTransposeFilter);
 
-  // To simplify logic, we only consider linalg ops with transposes who's
-  // outputs maps are identities
-  for (OpOperand *opOperand : linalgOp.getOutputOperands()) {
-    if (!linalgOp.getTiedIndexingMap(opOperand).isIdentity()) {
-      return failure();
-    }
-  }
-
-  // Determine which operands are transposed.
-  SmallVector<int64_t, 4> transposedOperandIndices;
-  for (auto indexMapPair : llvm::enumerate(linalgOp.getIndexingMapsArray())) {
-    if (isSharedMemTranspose(indexMapPair.value())) {
-      transposedOperandIndices.push_back(indexMapPair.index());
-    }
+  // Checks preconditions for shared mem transpose.
+  if (!opInfo.isTranspose() || opInfo.isDynamic() || opInfo.isReduction() ||
+      !opInfo.isTwoThreeLoops()) {
```

Comment:
yeah, I would use linalgOp for the simple checks here and move the `isTwoThreeLoops` check to this file since it's a specific check.